src/lib/Runner.ts

```diff
@@ -459,18 +456,51 @@ export class Runner {
         return this.next();
       case 'FAILED':
         logMessage('Moving landRequest to failed state');
-        await landRequest.setStatus('fail', 'Landkid build failed');
+        await this.failAllDependents(landRequestStatus, 'fail', 'Landkid build failed');
         return this.next();
       case 'STOPPED':
-        logMessage('Moving landRequest to aborted state');
-        await landRequest.setStatus('aborted', 'Landkid pipelines build was stopped');
+        logMessage(`Moving landRequest to aborted state`);
+        await this.failAllDependents(
+          landRequestStatus,
+          'aborted',
+          'Landkid pipelines build was stopped',
+        );
         return this.next();
       default:
         logMessage('Dont know what to do with build status, ignoring');
         break;
     }
   };
 
+  // Mark the current request as failed / aborted and fail all dependents within 'status-transition' lock.
+  // So, multiple free slots are available in one go for the speculation engine to process.
+  failAllDependents = async (
+    currentLandRequestStatus: LandRequestStatus,
+    currentRequestState: IStatusState,
+    currentRequestReason: string,
+  ) => {
+    await withLock(
+      'status-transition',
+      async (lockId: Date) => {
+        await currentLandRequestStatus.request.setStatus(currentRequestState, currentRequestReason);
+        const running = await this.getRunning();
+        const dependents = Runner.getDependents(running, currentLandRequestStatus);
+
+        Logger.info('Failing all dependents', {
+          namespace: 'lib:runner:failAllDependents',
+          landRequestId: currentLandRequestStatus.requestId,
+          pullRequestId: currentLandRequestStatus.request.pullRequestId,
+          dependentsPullRequestId: dependents.map((dependent) => dependent.request.pullRequestId),
+        });
+        for (const dependent of dependents) {
+          await this.failDueToDependency(dependent, [currentLandRequestStatus], lockId);
+        }
+      },
+      undefined,
+      100,
```

Comment:
I'm not sure that we should set a `ttl` here as the `next()` function itself doesn't have one and all the others that do have a `ttl` are separate interval functions that have a comment about not blocking `next`. I'd imagine we still want to be blocking `next` and so shouldn't set a 100ms `ttl`.

I don't quite understand the original reasoning for 100ms `ttl` anyway even after reading the comment, maybe @cheesehary recalls?

Reply:
Yes, it make sense for the interval functions to have lower `ttl`. In this case, we want to block the execution of the `next` functions.